packages/protocol/contracts/identity/FederatedAttestations.sol

```diff
@@ -0,0 +1,386 @@
+  function lookupAttestations(
+    bytes32 identifier,
+    address[] memory trustedIssuers,
+    uint256 maxAttestations
+  ) public view returns (address[] memory, uint256[] memory, address[] memory) {
+    // Cannot dynamically allocate an in-memory array
+    // For now require a max returned parameter to pre-allocate and then shrink
+    // TODO ASv2 is it a risk to allocate an array of size maxAttestations?
+    // Same index corresponds to same attestation
+    address[] memory accounts = new address[](maxAttestations);
+    uint256[] memory issuedOns = new uint256[](maxAttestations);
+    address[] memory signers = new address[](maxAttestations);
+
+    uint256 currIndex = 0;
+
+    for (uint256 i = 0; i < trustedIssuers.length; i = i.add(1)) {
+      uint256 numTrustedIssuers = identifierToAddresses[identifier][trustedIssuers[i]].length;
+      for (uint256 j = 0; j < numTrustedIssuers; j = j.add(1)) {
+        // Only create and push new attestation if we haven't hit max
+        if (currIndex < maxAttestations) {
+          // solhint-disable-next-line max-line-length
+          OwnershipAttestation memory attestation = identifierToAddresses[identifier][trustedIssuers[i]][j];
+          if (!revokedSigners[attestation.signer]) {
+            accounts[currIndex] = attestation.account;
+            issuedOns[currIndex] = attestation.issuedOn;
+            signers[currIndex] = attestation.signer;
+            currIndex = currIndex.add(1);
+          }
+        } else {
+          break;
+        }
+      }
+    }
+
+    // Trim returned structs if necessary
+    if (currIndex < maxAttestations) {
+      address[] memory trimmedAccounts = new address[](currIndex);
+      uint256[] memory trimmedIssuedOns = new uint256[](currIndex);
+      address[] memory trimmedSigners = new address[](currIndex);
+
+      for (uint256 i = 0; i < currIndex; i = i.add(1)) {
+        trimmedAccounts[i] = accounts[i];
+        trimmedIssuedOns[i] = issuedOns[i];
+        trimmedSigners[i] = signers[i];
+      }
+      return (trimmedAccounts, trimmedIssuedOns, trimmedSigners);
+    } else {
+      return (accounts, issuedOns, signers);
+    }
+  }
+
+  /**
+   * @notice Returns identifiers mapped (by unrevoked signers) to a
+   *   given account address by a list of trusted issuers
+   * @param account Address of the account
+   * @param trustedIssuers Array of n issuers whose identifier mappings will be used
+   * @param maxIdentifiers Limit the number of identifiers that will be returned
+   * @return Array (length <= maxIdentifiers) of identifiers
+   * @dev Adds identifier info to the arrays in order of provided trustedIssuers
+   * @dev Expectation that only one attestation exists per (identifier, issuer, account)
+   */
+
+  // TODO ASv2 consider also returning an array with counts of identifiers per issuer
+
+  function lookupIdentifiersByAddress(
+    address account,
+    address[] memory trustedIssuers,
+    uint256 maxIdentifiers
```

Comment:
Same as above, my guess would be to iterate through twice and use the first time to determine the appropriate length for the arrays so we can avoid needing to copy / trim the array data. We can still provide this parameter to avoid long and expensive transactions though